File: JS/Algorithm/FindoutMaximumProfitByDP.js
function maxProfit(price, n, k){
	// vector to store results of subproblems profit[i] stores maximum profit using at most i transactions up to the current day
	let profit = new Array(k + 1).fill(0);
	let prevDiff = new Array(k + 1).fill(Number.MIN_SAFE_INTEGER);
	// fill the table in bottom-up fashion
	for(let j = 0; j < n; j++){
		for(let i = 1; i <= k; i++){
			prevDiff[i] = Math.max(prevDiff[i], profit[i - 1] - price[j]);
			profit[i] = Math.max(profit[i], price[j] + prevDiff[i]);
		};
	};
	return profit[k];
};
// driver code
let k = 3;
let price = [12, 14, 17, 10, 14, 13, 12, 15];
let n = price.length;
console.log("Maximum profit is: " + maxProfit(price, n, k));
// contributed by phasing17

File: C#/Algorithm/FindoutMaximumProfitByDP.cs
// C# progarm to find out maximum profit by buying & selling a share atmost k times given stock price of n days
using System;
class H32 {
    static int maxProfit(int[] price, int n, int k){
        // table to store results of subproblems, ptrofit[t][i] stores maximum profit using atmost t transactions up to day i (including day i)
        int[, ] profit = new int[k+1, n+1];
        // for day 0, you can't arn money irrespective of how many times you trades
        for(int i = 0; i <= k; i++)
            profit[i, 0] = 0;
        // profit is 0 if we don't do any transactions
        for(int j = 0; j <= n; j++)
            profit[0, j] = 0;
        // fill the table in bottom-up fashion
        for(int i =1; i <= k; i++){
            int prevDiff = int.MinValue;
            for(int j = 1; j < n; j++){
                prevDiff = Math.Max(prevDiff, profit[i - 1, j - 1] - price[j - 1]);
                profit[i, j] = Math.Max(profit[i, j - 1], price[j] + prevDiff);
            };
        };
        return profit[k, n - 1];
    };
    // driver code to test above
    public static void Main(){
        int k = 3;
        int[] price = {12, 14, 17, 10, 14, 13, 12, 15};
        int n = price.Length;
        Console.Write("Maximum profit is: " + maxProfit(price, n, k));
    };
};
// Cre: Sam007 on GeeksforGeeks